### PR TITLE
DROOLS-3763: New API codegen prototype

### PR DESCRIPTION
**Pull request by evacchi**
API we discussed together:

- `RuleUnit<T>/RuleUnitInstance<T>`
- one KieSession per RuleUnitInstance
- generated Module class with  code-generated `create${T}RuleUnit()`
- refactored Maven plugin

see examples for reference https://github.com/kiegroup/submarine-examples/pull/5

### CODE REVIEW

**Comment by lucamolteni** [comment]
Was the "Simple" untyped `<Object>` Unit implemented?

**Comment by evacchi** [comment]
not yet, I'll add it soon. We'll have to think a little bit about implication (if it's library-based you can't "discover it" as source code but look in classpath etc etc)

File: drools/drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/ModelSourceClass.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    public void write(MemoryFileSystem srcMfs) {
+        srcMfs.write(CanonicalModelKieProject.PROJECT_MODEL_SOURCE, generate().getBytes());
+    }
+
+    public String generate() {
+        StringBuilder sb = new StringBuilder();
+        sb.append(
+                "package org.drools.project.model;\n" +
+                        "\n" +
```
[the remark this answers is not in the document]
Comment (evacchi) [reply drools/drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/ModelSourceClass.java]: this is code we are going to drop, I only refactored it from the plugin. So... yes, by dropping it altogether

[the remark this answers is not in the document]
Comment (lucamolteni) [reply drools/drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/builder/ModelSourceClass.java]: Drop all the code!

**Comment by mbiarnes** [comment]
Jenkins please retest this

**Comment by mbiarnes** [comment]
Jenkins retest this please

**Comment by evacchi** [comment]
after discussion with @mariofusco I am merging this.

**Pull request merged**